### PR TITLE
Add `app.hide()` and `app.show()`

### LINKED ISSUE #3694: API to hide and show app
I need ability to programmatically hide the app as if the user did <kbd>Cmd</kbd> <kbd>H</kbd>.

I'm thinking: `app.show()` and `app.hide()`.

### PR DESCRIPTION
**Pull request by KevinMartin**
Fixes #3694

### CODE REVIEW

**Comment by KevinMartin** [comment]
@sindresorhus Is this kind of what you're looking for?

I made the `app.show()` function have an option `focus` parameter in case you just want to bring the windows to the background (assuming there is another active program)

**Comment by zcbenz** [comment]
I think we can simply mark the API as OS X only and only implement it there. Correctly implement it for other platform is hard and meaningless to me.

**Comment by KevinMartin** [comment]
What's wrong with the non-OSX code?

File: atom/browser/api/lib/app.coffee [hunk 1]
```diff
@@ -34,6 +34,27 @@ app.setAppPath = (path) ->
...
+  else
+    BrowserWindow.getAllWindows().forEach window ->
+      window._wasMinimized = window.isMinimized()
+      window._wasFocused = window.isFocused()
+
+      window.hide()
+
+app.show = (focus = false) ->
+  if process.platform is 'darwin'
+    action = focus ? 'activateIgnoringOtherApps:' : 'unhide:'
```
Comment (zcbenz) [inline atom/browser/api/lib/app.coffee]: The `activateIgnoringOtherApps` was already implemented in the `Browser::Focus` native code, if you really want to expose it you can map `app.focus` to it.

**Comment by zcbenz** [comment]
> What's wrong with the non-OSX code?

At least on Window, it doesn't restore the minimized and fullscreen state of windows, and gives focus to wrong windows. (calling `show()` would focus a window)

If you really want to have fallback code on other platforms, calling `hide()` and `showInactive()` is enough. Users who want other behaviors should write their own code.

**Comment by paulcbetts** [comment]
The concept of apps "Hiding" on Windows or Linux is pretty weird tbh, I'd also rather this be a no-op on Windows / Linux to encourage app developers not to use it

**Comment by sindresorhus** [comment]
> I'd also be :+1: on not providing this as an explicit method at all, just having OS X's windows implicitly handle Cmd-H

That's how it currently works. I need this method as I want the same behavior of Cmd-H when the window is closed (because I need to keep it in the background).

> perhaps emitting an event that allows an app to cancel it

I guess that could be useful too, but it doesn't solve my problem.

**Comment by zcbenz** [comment]
I think we should only add new APIs when:

1. a job can not be done with existing APIs, like `Menu.sendActionToFirstResponder`,
2. there is a common pattern for certain jobs, like `Menu.buildFromTemplate`.

Hiding an app doesn't belong to either condition, it can be done with existing APIs, and on Windows and Linux it is a rather rare task.

The only reason I think we can add `app.show/hide` APIs on OS X is, hiding/showing an app is kind of common task on this platform.

**Comment by sindresorhus** [comment]
> The only reason I think we can add app.show/hide APIs on OS X is, hiding/showing an app is kind of common task on this platform.

:+1:

**Comment by KevinMartin** [comment]
So, should I close this PR then, @zcbenz & @paulcbetts ?

**Comment by zcbenz** [comment]
> So, should I close this PR then, @zcbenz & @paulcbetts ?

I think you can change `app.show` and `app.hide` to simple wrappers of `Menu.sendActionToFirstResponder('unhide')` and `Menu.sendActionToFirstResponder('hide')`.

**Comment by sindresorhus** [comment]
> I think you can change app.show and app.hide to simple wrappers of Menu.sendActionToFirstResponder('unhide') and Menu.sendActionToFirstResponder('hide').

:+1:

**Comment by zcbenz** [comment]
I'm closing this because there hasn't been any update since last code review, will reopen when there is update.

**Pull request closed without merging**